### DISCUSSION TITLE
Debugging custom components

### DISCUSSION THREAD

**Discussion by me-nkr** [hneemann/Digital#1312] [Q&A]
I am trying to create a custom component and debug it, and there are two issues I am facing
- when debugging the component is greyed out 
![image](https://github.com/user-attachments/assets/89eba3a8-447a-488f-92a0-451a67db13db)
  - I tried using `java -cp "digitaln2t-1.0-SNAPSHOT.jar:/path/to/Digital/Digital.jar" com.github.itoshkov.digitaln2t.Nand2TetrisComponentSource`, `java -cp "target/pluginExample-1.0-SNAPSHOT.jar:Digital.jar" de.neemann.digital.plugin.DemoCompone
ntSource` and tried using vscode java debugger, the same is the result
  - it worked when I provided the jar path as Java Library option in Digital Settings
 
- this is a minor inconvenience, but just wanted to know about it, when using vscode java debugger and making changes to the file the changes are not reflected in the Digital instance that was opened with debugger.

**Comment by hneemann** [comment]
This happens if there are two components with the same name. The simulator identifies the components by name. If there are two with the same name, this name is no longer unique. For example, has a jar file been included which contains a component with a certain name and another component with the same name registered in debug mode?

**Reply by me-nkr** [reply]
- I freshly cloned the demo repo
- replaced Digital.jar with what I use
- inside the repo, I did `mvn install`
- then `java -cp "target/pluginExample-1.0-SNAPSHOT.jar:Digital.jar" de.neemann.digital.plugin.DemoComponentSource`
- Digital was opened but the custom components were greyed out

And to answer your question, I don't think there is.

**Comment by me-nkr** [comment]
Some tinkering later, here are my observations
- I did get it working in an accident, I opened a dig file which has some custom dig components in the same directory
  - now the Nand2Tetris component can be selected and custom components also can be selected but custom components has no icon preview until it's selected once
- I do feel like this has something to do with [these lines of code](https://github.com/hneemann/Digital/blob/426a1e11700c8a6e3dfd96a49a1fdba5389ec05d/src/main/java/de/neemann/digital/draw/library/ElementLibrary.java#L697-L699), but I am not sure since the custom components made it work

will update if I get more info into the situation

**Comment by me-nkr** [comment]
Hey @hneemann, I found out why it's not working and why it worked when there was a file in file history.

It has to do with the order of executions and the following lines of code
https://github.com/hneemann/Digital/blob/426a1e11700c8a6e3dfd96a49a1fdba5389ec05d/src/main/java/de/neemann/digital/gui/Main.java#L194-L202

# Here is a detailed explanation
## The order of executions is as follows when debugging
  1. create a new element library
    a. calls the  `LibraryNode.traverse()` method from `populateNodeMap()`
    b. calls the `LibraryNode.traverse()` method directly 
https://github.com/hneemann/Digital/blob/426a1e11700c8a6e3dfd96a49a1fdba5389ec05d/src/main/java/de/neemann/digital/draw/library/ElementLibrary.java#L255-L263
  2. register components from custom Java components to the above-created library using the `ElementLibrary.registerComponentSource()` method
  3. set the above created and modified library as the library of `MainBuilder` using the `MainBuilder.setLibrary()` method
  4. then the `MainBuilder.openLater()` method calls the `MainBuilder.build()` method which in turn instantiates the `Main` class
  5. while instantiating the `Main` class, as per the first block of code linked, the `LibraryNode.traverse()` method is called once more as a result of calling the `Main.loadFile()` method which only happens if there is a file in File History
## The gotcha
First two calls to the `LibraryNode.traverse()` method happens before the `ElementLibrary.registerComponentSource()` and  the `MainBuilder.setLibrary()` method calls. Because of this, the custom Java components are not in the element library when the `LibraryNode.traverse()` method is called which is responsible for marking library nodes as unique. Since the custom Java components are not marked as unique, they become unusable. And since the `LibraryNode.traverse()` method is called once more after all this when there is a file in File History, the custom Java components become usable.
